File: backend/scoring/scorer.py
from dataclasses import dataclass, field
from typing import Dict

from backend.config import TECHNICAL_WEIGHTS
from backend.utils.logger import setup_logger

logger = setup_logger("scorer")


@dataclass
class Signal:
    signal: str = "HOLD"
    confidence: int = 0
    composite_score: float = 0.0
    breakdown: Dict = field(default_factory=dict)
    risk_levels: Dict = field(default_factory=dict)
    filter_reasons: list = field(default_factory=list)


class ScoringEngine:
    def __init__(self):
        self.weights = TECHNICAL_WEIGHTS

    def calculate(
        self,
        technical_score: float,
        onchain_score: float,
        sentiment_score: float,
        fundamental_score: float,
        risk_score: float,
        current_price: float = 0,
        atr: float = 0,
    ) -> Signal:
        t = technical_score / 10.0
        o = onchain_score / 10.0
        s = sentiment_score / 10.0
        f = fundamental_score / 10.0
        r = risk_score / 10.0

        composite = (
            t * self.weights["technical"]
            + o * self.weights["onchain"]
            + s * self.weights["sentiment"]
            + f * self.weights["fundamental"]
            + r * self.weights["risk"]
        )

        factors = [t, o, s, f]
        mean = sum(factors) / len(factors) if factors else 0
        variance = sum((x - mean) ** 2 for x in factors) / len(factors) if factors else 0
        agreement = 1 - min(variance ** 0.5, 1.0)

        if composite > 0.5:
            signal_type = "STRONG_BUY"
        elif composite > 0.2:
            signal_type = "BUY"
        elif composite > -0.2:
            signal_type = "HOLD"
        elif composite > -0.5:
            signal_type = "SELL"
        else:
            signal_type = "STRONG_SELL"

        confidence = min(100, max(0, int(
            abs(composite) * 60
            + agreement * 30
            + 10
        )))

        risk_levels = {}
        if current_price > 0 and atr > 0:
            if signal_type in ("BUY", "STRONG_BUY"):
                stop_loss = current_price - (2 * atr)
                take_profit = current_price + (3 * atr)
            elif signal_type in ("SELL", "STRONG_SELL"):
                stop_loss = current_price + (2 * atr)
                take_profit = current_price - (3 * atr)
            else:
                stop_loss = 0
                take_profit = 0

            risk = abs(current_price - stop_loss) if stop_loss > 0 else 1
            reward = abs(take_profit - current_price) if take_profit > 0 else 0
            rr_ratio = reward / risk if risk > 0 else 0

            risk_levels = {
                "entry": current_price,
                "stop_loss": stop_loss,
                "take_profit": take_profit,
                "rr_ratio": round(rr_ratio, 2),
                "risk_pct": round(risk / current_price * 100, 2) if current_price > 0 else 0,
                "reward_pct": round(reward / current_price * 100, 2) if current_price > 0 else 0,
            }

        return Signal(
            signal=signal_type,
            confidence=confidence,
            composite_score=round(composite, 4),
            breakdown={
                "technical": round(technical_score, 2),
                "onchain": round(onchain_score, 2),
                "sentiment": round(sentiment_score, 2),
                "fundamental": round(fundamental_score, 2),
                "risk": round(risk_score, 2),
            },
            risk_levels=risk_levels,
        )

File: cf-worker/src/ml/scorer.ts
import { Signal } from '../types';
import { CONFIG } from '../types/config';

export function calculateSignal(
  technicalScore: number,
  onchainScore: number,
  sentimentScore: number,
  fundamentalScore: number,
  riskScore: number,
  currentPrice: number,
  atr: number,
): Signal {
  const t = technicalScore / 10;
  const o = onchainScore / 10;
  const s = sentimentScore / 10;
  const f = fundamentalScore / 10;
  const r = riskScore / 10;

  const composite =
    t * CONFIG.TECHNICAL_WEIGHTS.technical +
    o * CONFIG.TECHNICAL_WEIGHTS.onchain +
    s * CONFIG.TECHNICAL_WEIGHTS.sentiment +
    f * CONFIG.TECHNICAL_WEIGHTS.fundamental +
    r * CONFIG.TECHNICAL_WEIGHTS.risk;

  const factors = [t, o, s, f];
  const mean = factors.reduce((a, b) => a + b, 0) / factors.length;
  const variance = factors.reduce((a, b) => a + (b - mean) ** 2, 0) / factors.length;
  const agreement = 1 - Math.min(Math.sqrt(variance), 1);

  let signalType: Signal['signal'] = 'HOLD';
  if (composite > 0.5) signalType = 'STRONG_BUY';
  else if (composite > 0.2) signalType = 'BUY';
  else if (composite > -0.2) signalType = 'HOLD';
  else if (composite > -0.5) signalType = 'SELL';
  else signalType = 'STRONG_SELL';

  const confidence = Math.min(100, Math.max(0, Math.round(
    Math.abs(composite) * 60 + agreement * 30 + 10,
  )));

  let stopLoss = 0, takeProfit = 0, rrRatio = 0, riskPct = 0, rewardPct = 0;
  if (currentPrice > 0 && atr > 0 && signalType !== 'HOLD') {
    const direction = signalType.includes('BUY') ? 1 : -1;
    stopLoss = currentPrice - direction * 2 * atr;
    takeProfit = currentPrice + direction * 3 * atr;
    const risk = Math.abs(currentPrice - stopLoss);
    const reward = Math.abs(takeProfit - currentPrice);
    rrRatio = Math.round((reward / risk) * 100) / 100;
    riskPct = Math.round((risk / currentPrice) * 10000) / 100;
    rewardPct = Math.round((reward / currentPrice) * 10000) / 100;
  }

  return {
    signal: signalType,
    confidence,
    composite_score: Math.round(composite * 10000) / 10000,
    price: currentPrice,
    breakdown: {
      technical: Math.round(technicalScore * 100) / 100,
      onchain: Math.round(onchainScore * 100) / 100,
      sentiment: Math.round(sentimentScore * 100) / 100,
      fundamental: Math.round(fundamentalScore * 100) / 100,
      risk: Math.round(riskScore * 100) / 100,
    },
    risk_levels: { entry: currentPrice, stop_loss: stopLoss, take_profit: takeProfit, rr_ratio: rrRatio, risk_pct: riskPct, reward_pct: rewardPct },
    filter_reasons: [],
    technical: { raw: technicalScore, normalized: technicalScore, indicators: {} },
    onchain: { raw: onchainScore, normalized: onchainScore, metrics: {} },
    sentiment: { score: sentimentScore, classification: 'NEUTRAL', breakdown: {} },
    fear_greed: 50,
    time: new Date().toISOString(),
  };
}

export function applyFilters(signalData: Signal, marketData: Record<string, any>): Signal {
  let confidence = signalData.confidence;
  const reasons: string[] = [];

  if (signalData.signal === 'HOLD') return signalData;

  const high24h = marketData.high_24h || 0;
  const low24h = marketData.low_24h || 0;
  const price = marketData.price || 1;
  const range24h = price > 0 ? (high24h - low24h) / price : 0;
  if (range24h < 0.02) { confidence -= 30; reasons.push('RANGE_BOUND'); }

  const volRatio = marketData.volume_ratio || 1;
  if (volRatio < 0.5) { confidence -= 25; reasons.push('LOW_VOLUME'); }

  if (marketData.fear_greed > 90 || marketData.fear_greed < 10) {
    confidence -= 20;
    reasons.push('EXTREME_SENTIMENT');
  }

  signalData.confidence = Math.max(0, confidence);
  signalData.filter_reasons = reasons;

  if (signalData.confidence < 30) {
    signalData.signal = 'HOLD';
    signalData.confidence = 0;
  }

  return signalData;
}

export function validateRisk(signalData: Signal): Signal {
  const rr = signalData.risk_levels.rr_ratio;
  if (rr < CONFIG.MIN_RR_RATIO && signalData.signal !== 'HOLD') {
    signalData.confidence = Math.max(0, signalData.confidence - 20);
    signalData.filter_reasons.push(`LOW_RR_${rr}`);
  }
  if (signalData.risk_levels.risk_pct > 5) {
    signalData.confidence = Math.max(0, signalData.confidence - 15);
    signalData.filter_reasons.push(`HIGH_RISK_${signalData.risk_levels.risk_pct}%`);
  }
  return signalData;
}
